python/src/aac/execute/aac_execution_result.py

```diff
@@ -0,0 +1,102 @@
+
+    SUCCESS = 0
+    CONSTRAINT_FAILURE = auto()
+    CONSTRAINT_WARNING = auto()
+    PARSER_FAILURE = auto()
+    PLUGIN_FAILURE = auto()
+    OPERATION_CANCELLED = auto()
+    GENERAL_FAILURE = auto()
+
+
+class MessageLevel(Enum):
+    """An enumeration that represents the level of a message."""
+
+    DEBUG = auto()
+    INFO = auto()
+    WARNING = auto()
+    ERROR = auto()
+
+
+@attrs(slots=True, auto_attribs=True)
+class ExecutionMessage:
+    """Provides a message for the user."""
+
+    message: str = attrib(validator=validators.instance_of(str))
+    level: MessageLevel = attrib(validator=validators.instance_of(MessageLevel))
+    source: Optional[AaCFile] = attrib(
+        validator=validators.optional(validators.instance_of(AaCFile))
+    )
+    location: Optional[SourceLocation] = attrib(
+        validator=validators.optional(validators.instance_of(SourceLocation))
+    )
+
+
+@attrs(slots=True, auto_attribs=True)
+class ExecutionResult:
+    """Provides information regarding the results of the execution of a plugin.
+
+    Attributes:
+        name (str): The name of the plugin whose results are included.
+        status_code (ExecutionStatus): A status code for the plugin execution.
+        messages (list[str]): A list of messages for the user.
+    """
+
+    plugin_name: str = attrib(validator=validators.instance_of(str))
+    plugin_command_name: str = attrib(validator=validators.instance_of(str))
+    status_code: ExecutionStatus = attrib(
+        validator=validators.instance_of(ExecutionStatus)
+    )
+    messages: list[ExecutionMessage] = attrib(
+        default=Factory(list), validator=validators.instance_of(list)
+    )
+
+    def add_message(self, message: ExecutionMessage) -> None:
+        """Add a message to the list of messages."""
+        self.messages.append(message)
+
+    def add_messages(self, messages: list[ExecutionMessage]) -> None:
+        """Add messages to the list of messages."""
+        self.messages.extend(messages)
+
+    def is_success(self) -> bool:
+        """Return True if the command succeeded; False, otherwise."""
+        return self.status_code == ExecutionStatus.SUCCESS
+
+    def get_messages_as_string(self) -> str:
+        """Return the output messages as a combined string."""
+        result = ""
+        for message in self.messages:
+            result += message.message + linesep
+            if message.source is not None:
+                result += f"  Source: {message.source.uri}"
+                if message.location is not None:
+                    result += f" ({message.location.line}:{message.location.column}:{message.location.position}:{message.location.span})"
+                result += linesep
+        return result
+
+
+@attrs(slots=True)
+class ExecutionError(Exception):
+    """A base class representing a plugin error condition."""
+
+    message: str = attrib(validator=validators.instance_of(str))
```

Comment:
I'm not sure what you mean with this comment.

Reply:
The `LanguageError` exception class includes attributes for `message` and `location` about the encountered error. Why would we not want those same pieces of information for at least `ExecutionError`? Could see how it might be a bit pointless for the `OperationCancelled`, but then you could provide back to the user where within execution things got stopped or errored out. 

`ExecutionMessage` has a `location` attribute as well, so having the custom exceptions all provide location information, it would be used in the returned information to the user via `ExecutionResult`.

Reply:
OK...makes sense now.  I like the idea and recommend we make this change after we figure out how to reliably get that location data from lexemes.  When I look at the existing uses, it's clear that this location data must be optional.  Let's write a story to do this in conjunction with or after improvements to "location determination".